### PR TITLE
ProfileBuilder Servlet

### PR DESCRIPTION
**Pull request by samuelsandoval1**
## Proposed Changes
Created the ProfileBuilderServlet which includes a POST method that posts the mentee and mentor form information to the Datastore. Also added testts in the ProfileBuilderServletTest file.

## Types of Changes
What types of changes does your code introduce to Techish Project?
Put an x in the boxes that apply

- [ ]  Bugfix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Documentation Update (if none of the other choices apply)

## Checklist
Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.

- [x]  I have added tests that prove my fix is effective or that my feature works
- [x]  I have added necessary documentation (if appropriate)
- [x]  Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+public class ProfileBuilderServletTest {
+    
+  @Mock
+  private HttpServletRequest request;
+
+  @Mock
+  private HttpServletResponse response;
+
+
+  private ProfileBuilderServlet userServlet;
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:55]: For consistency, could you either name the servlet UserServlet or change the variable to be profileBuilderServlet?

If this won't be modifed, consider using the final keyword with the ProfileBuilderServlet.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    when(request.getParameter("name-input")).thenReturn(menteeName);
+    String school = "University of California Berkeley";
+    when(request.getParameter("school-input")).thenReturn(school);
+    String major = "Computer Science";
+    when(request.getParameter("major-input")).thenReturn(major);
+  }
+
+  @Test
+  public void doPostRedirectsUrl() throws IOException, ServletException {
+    addMentorProfile();
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: Is there a reason we need to call addMentorProfile() for testing that a POST request will send a redirect?

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    userServlet.doPost(request, response);
+
+    verify(response, times(1)).sendRedirect("/logged_in_homepage.html");
+   }
+  
+  @Test
+  public void addMentorProfileThenCheckSize() throws IOException, ServletException {
+    DatastoreService dataService = DatastoreServiceFactory.getDatastoreService();
+    addMentorProfile();
+    userServlet.doPost(request, response);
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:109]: Could you add additional tests to ensure your servlet works as expected with multiple or zero mentor/mentees?

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    when(request.getParameter("school-input")).thenReturn(school);
+    String major = "Computer Science";
+    when(request.getParameter("major-input")).thenReturn(major);
+
+    userServlet.doPost(request, response);
+    Filter schoolFilter = new FilterPredicate("school", FilterOperator.EQUAL, school);
+    PreparedQuery results = ds.prepare(new Query("User").setFilter(schoolFilter));
+    String result = "";
+    for (Entity entity : results.asIterable()) {
+        result = (String) entity.getProperty("school");
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:140]: rather than checking the fields in separate tests, consider verifying that the entity that is returned from datastore matches the entity you expect to be added based on your request parameters.

File: src/main/webapp/ts/profile.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+async function getProfile() {
+    const response = await fetch('/profile');
```
Comment (jerrychangusa) [inline src/main/webapp/ts/profile.ts:2]: consider separating the frontend changes for a separate pull request as this makes this change more focused and eaiser to review.

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PersistentUserRepository emptyRepo = new PersistentUserRepository();
+    String userType = request.getParameter("user-Type");
+    
+    if(userType.equals("Mentee")){
+        String userName = request.getParameter("name-input");
+        User currentUser = new User(userName);
+
```
Comment (jerrychangusa) [inline src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:36]: maybe add a TODO for adding the Google Id?

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for creating new profiles. */
+@WebServlet("/profile")
+public class ProfileBuilderServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PersistentUserRepository emptyRepo = new PersistentUserRepository();
```
Comment (jerrychangusa) [inline src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:30]: consider moving the PersistentUserRepository to class variable rather than creating a new object for each new POST request.

File: src/main/java/com/google/sps/data/User.java [hunk 8]
```diff
@@ -34,10 +34,10 @@ public String toString() {
   } 
 
   public ProfileType toEnum(String input) {
-      if(input == "student" || input == "STUDENT") {
-          input = "STUDENT";
+      if(input == "Mentee" || input == "MENTEE") {
```
Comment (ghandre) [inline src/main/java/com/google/sps/data/User.java:37]: nit: [L37 and L40]: Maybe input.equalsIgnoreCase("STUDENT") ? More succinct and you also catch "Student".

File: src/main/java/com/google/sps/data/User.java [hunk 9]
```diff
@@ -65,8 +65,8 @@ public void addSpecialty(String toAdd) {
   }
 
   public void setProfileType(ProfileType input) {
-      if(input == ProfileType.STUDENT) {
-        this.userType = ProfileType.STUDENT;
+      if(input == ProfileType.MENTEE) {
```
Comment (ghandre) [inline src/main/java/com/google/sps/data/User.java:68]: nit: Slightly more succinct as
`if (input == ProfileType.MENTOR || input == ProfileType.MENTEE) {
  this.userType = input;
} else { ...`

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        String userName = request.getParameter("name-input");
+        User currentUser = new User(userName);
+
+        String userSchool = request.getParameter("school-input");
+        currentUser.setSchool(userSchool);
+        String userMajor = request.getParameter("major-input");
+        currentUser.setMajor(userMajor);
+        emptyRepo.addUser(currentUser);
+    }
+    else if(userType.contains("Mentor")){
```
Comment (ghandre) [inline src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:43]: L43: equals instead of contains?

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 7 above)
Comment (ghandre) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:30]: +1
Also, emptyRepo as a name isn't necessarily accurate, no? Maybe just 'repository'.

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+        String userCompany = request.getParameter("company-input");
+        currentUser.setCompany(userCompany);
+        String userSpecialty = request.getParameter("specialty-input");
+        currentUser.addSpecialty(userSpecialty);
+
+        emptyRepo.addUser(currentUser);
+    }
+    else{
+        throw new IOException("Error invalid usertype");
```
Comment (ghandre) [inline src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:55]: [L55] Maybe append the offending usertype to the debug message so it's more useful?

File: src/main/webapp/ts/profile.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+    const data = await response.json();
+    console.log(data);
```
Comment (ghandre) [inline src/main/webapp/ts/profile.ts:4]: [L4] Debug artifact?

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+    localHelper.setUp();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    localHelper.tearDown();
+  }
+    
+  public void addMentorProfile() {
```
Comment (ghandre) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:75]: 'addMentorProfile' sounds like you're adding it to the repository; Maybe 'setRequestAsMentor' or something of the like? Similarly for addMenteeProfile.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 1 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: I just call it to add the fake user data so that the post request posts the user information to datastore. Without it the post method cannot properly function and I get a null pointer exception.

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: The test pretty much checks to see that the doPost posts the user information to datastore then redirects at the end.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 3 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Made test that adds Mentee, post. Then add Mentor, and post

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/main/webapp/ts/profile.ts (hunk 5 above)
Comment (samuelsandoval1) [reply src/main/webapp/ts/profile.ts:2]: Ok

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (jerrychangusa) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: I read that this test verifies that posting user information to the servlet causes a redirect. 

Could you also include a test where a POST request has missing user information? We should ensure that the empty request case is well tested to prevent the servlet from throwing a null pointer exception in production.

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 6 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:37]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: For this additional test, would doing this check be sufficient

1. Add mentor profile with a missing information : specialty in this case
2. call Post request
3. do this assert test 
`    Assert.assertEquals(1, dataService.prepare(new Query("User")).countEntities(withLimit(10)));
`

This test returns true.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 4 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:152]: So, you'd like to have one test check that the entity being return matches the parameters? All in one test?

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 7 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:30]: Would you like the PersistentUserRepository class variable to be just a variable, or also have it create an object when its initialized as a class variable. 

Ex. 
`
  private static PersistentUserRepository repository = 
    new PersistentUserRepository();`

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (jerrychangusa) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: Testing with one empty field field (e.g. specialty) provides some coverage. I am more concerned about the case where the user makes a POST request without including _any_ fields. You can write this test using the following guideline:

1.  Create a mock request where all accessed fields return a nullptr.
2. Call the POST method with the mock request.
3. Verify that the servlet returns a redirect and does not store any new data into datastore.

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: For step 3, is that 2 checks

like the verify it redirects

and the assert equals

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 4 above)
Comment (jerrychangusa) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:152]: Yes I am suggesting that you check multiple fields in the values from the datastore. If you set ten properties in your request, then you can verify that the ten properties exist in the datastore. 

The principle of separating tests so that each test verifies a single behavior causes your tests to be easier for others to understand. However to test a single behavior, you may need to validate that multiple conditions are true. Because we are testing whether a POST request will save an entity into the datastore, we should also check that all the fields of the entity are correct.

Let me know if that helps.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: Also, for step 1
shoultd it return null or nullptr

Nullptr said no symbol found


Then with null that just led to a nullptr exception error

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 4 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:152]: Yea that helps. I made a test to fetch from Datastore, get all the entity properties then compare to whats originally posted

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 14]
```diff
@@ -41,20 +41,27 @@ public PersistentUserRepository() {
   }
 
   public void addUserToDatabase(User user) {
+      // if for mentee, mentor
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java]: Done

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 15]
```diff
@@ -41,20 +41,27 @@ public PersistentUserRepository() {
...
-        userEntity.setProperty("specialties", specialties);
+
+    String school = user.getSchool();
+    String major = user.getMajor();
+    String company = user.getCompany();
+    Collection<String> specialties = user.getSpecialties();
+
+    if(school != null && major != null) {
+        userEntity.setProperty("school", school);  
+        userEntity.setProperty("major", major);
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:56]: Ill seperate the conditions individually instead, so one if per user property instead of combining school and major

[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:58]: Done

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 16]
```diff
@@ -41,20 +41,27 @@ public PersistentUserRepository() {
...
 
     Entity userEntity = new Entity("User");
     userEntity.setProperty("name", name);
-    
-    if(!specialties.isEmpty()) {
-        userEntity.setProperty("specialties", specialties);
+
+    String school = user.getSchool();
+    String major = user.getMajor();
+    String company = user.getCompany();
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:51]: I can't do ID yet because Hadley changed the ID type.

File: src/main/java/com/google/sps/data/User.java (hunk 8 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/User.java:37]: It won't be STUDENT ever because the usertype coming from the HTML is Mentee or Mentor

File: src/main/java/com/google/sps/data/User.java (hunk 9 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/User.java]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 7 above)
Comment (jerrychangusa) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:30]: Let's avoid creating the object multiple times by initializing the PersistentUserRepository just once. Your example should work.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (jerrychangusa) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: Sorry for the confusion, please use null instead of nullptr. If there is a nullptr exception error, your servlet may not be handling this case correctly.

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 7 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Ok. Implemented simpleton pattern for this

Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:99]: Ok

File: src/main/java/com/google/sps/data/PersistentUserRepository.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:58]: Added some tests

[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:58]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 2 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:100]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 4 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 13 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/main/java/com/google/sps/data/User.java (hunk 8 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/User.java]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 10 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 11 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 17]
```diff
@@ -41,20 +41,27 @@ public PersistentUserRepository() {
   }
 
   public void addUserToDatabase(User user) {
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:50]: Ok

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+    when(request.getParameter("user-Type")).thenReturn(userType);
+    String menteeName = "Sandy Cane";
+    when(request.getParameter("name-input")).thenReturn(menteeName);
+    String school = "Stanford University";
+    when(request.getParameter("school-input")).thenReturn(school);
+    String major = "Computer Science";
+    when(request.getParameter("major-input")).thenReturn(major);
+
+    profileBuilderServlet.doPost(request, response);
+    // Filter schoolFilter = new FilterPredicate("school", FilterOperator.EQUAL, school);
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:167]: Can you remove the commented out code on line 167?

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+  public void addMenteeProfileThenCheckSize() throws IOException, ServletException {
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    setRequestAsMentee();
+    profileBuilderServlet.doPost(request, response);
+
+    Assert.assertEquals(1, ds.prepare(new Query("User")).countEntities(withLimit(10)));
+  }
+
+  @Test
+  public void checkMenteeInfoIsSame() throws IOException, ServletException {
```
Comment (jerrychangusa) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:155]: Can you add a similar test for checking if the mentor info is the same as what is stored in data store? I see that you have a test for checking the mentor specialty. Can we check the other entity fields as well?

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userType = request.getParameter("user-Type");
+
+    if(userType != null) {
+        if(userType.equals("Mentee")){
+            //TODO: add the GoogleUserID
+            String userName = request.getParameter("name-input");
+            User currentUser = new User(userName);
```
Comment (jerrychangusa) [inline src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java:37]: If userName is null, then the rest of the code in this if block is not used. Consider moving the userName != null check before line 37 to exit this if block early when userName is null.

You can apply the same change to Line 54 for the Mentor else if block.

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 18 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 19 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:155]: Yea

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:155]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 20 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+    
+    Assert.assertEquals(mentorName, nameResult);
+    Assert.assertEquals(company, companyResult);
+    Assert.assertEquals(occupation, occupationResult);
+    Assert.assertArrayEquals(specialties, specialtiesResult);
+  }
+    
+    
+  @Test
+  public void checkMentorSpecialityIsSame() throws IOException, ServletException {
```
Comment (ghandre) [inline src/test/java/com/google/sps/ProfileBuilderServletTest.java:220]: [L220] Is this ('checkMentorSpecialityIsSame' method) redundant now that you have 'checkMentorInfoIsSame', which includes specialties? Or is there some other part of the logic you intended to test with it?
Same question for 'sendUserCheckIfNameMatches'.

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 22]
```diff
@@ -95,14 +137,33 @@ public PreparedQuery getQueryFilterForName(String name) {
...
-
     return userProfiles;
   }
+  
+  // function that fetches a single user, collection of users, filter on that user for id
+  public User fetchUserWithId(String userId) {
+    PreparedQuery results = getQueryFilterForId(userId);
+    Collection<User> userProfiles = fetchUserEntities(results);
+    if(userProfiles.size() == 1){
+        for (User userProfile : userProfiles) {
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+                if(userSpecialties != null) {
+                    for(String specialty: userSpecialties) {
+                        currentUser.addSpecialty(specialty);
+                    }
+                }
+                PersistentUserRepository.getInstance().addUser(currentUser);
+            } 
+        }
+        else{
+            throw new IOException("Error invalid usertype"+ userType);
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/main/webapp/profileBuilder.html [hunk 24]
```diff
@@ -77,26 +77,26 @@ <h1 class="headsection-heading mb-0">Profile Builder</h1>
...
-                            <input type="text" name="careerTitle-input" class="form-control"  placeholder="Career Title">
+                            <input type="text" name="careerTitle-input" class="form-control"  placeholder="Career Title" required>
                         </label>
                     </div>
                     
                     <!-- Specialty Input  -->
                     <label style= "color: black;"> Specialty:
                         <br>
-                        <select placeholder={{specialty}} name="specialty-Input" multiple>
+                        <select placeholder={{specialty}} name="specialty-input" Size="9"required>
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/webapp/profileBuilder.html]: Done

File: src/main/java/com/google/sps/data/PersistentUserRepository.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java]: Done

File: src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/servlets/ProfileBuilderServlet.java]: Done

File: src/main/webapp/profileBuilder.html (hunk 24 above)
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/main/webapp/profileBuilder.html]: Done

File: src/test/java/com/google/sps/PersistentUserRepositoryTest.java [hunk 25]
```diff
@@ -97,21 +98,74 @@ public void userSpecialtyWrittenBack() {
...
-    Assert.assertEquals(2, result);
+
+    Collection<String> expected = new HashSet<>();
+    expected.add("Front-end Development");
+    expected.add("UI / UX Design");
+    
+    Collection<User> allUsers = emptyRepo.getAllUsers();
+    Collection<String> specialtiesResult = new HashSet<>();
+
+    for (User allUser : allUsers) {
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/PersistentUserRepositoryTest.java]: Done

File: src/test/java/com/google/sps/PersistentUserRepositoryTest.java [hunk 26]
```diff
@@ -97,21 +98,74 @@ public void userSpecialtyWrittenBack() {
...
     userA.addSpecialty("Front-end Development");
     emptyRepo.addUser(userA);
-    int result = userA.getSpecialties().size();
-    Assert.assertEquals(2, result);
+
+    Collection<String> expected = new HashSet<>();
+    expected.add("Front-end Development");
+    expected.add("UI / UX Design");
+    
+    Collection<User> allUsers = emptyRepo.getAllUsers();
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/PersistentUserRepositoryTest.java]: Done

File: src/test/java/com/google/sps/PersistentUserRepositoryTest.java [hunk 27]
```diff
@@ -97,21 +98,74 @@ public void userSpecialtyWrittenBack() {
...
+    Collection<String> specialtiesResult = new HashSet<>();
+
+    for (User allUser : allUsers) {
+        int len = allUser.getSpecialties().size();
+        if(len > 0){
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/PersistentUserRepositoryTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 28]
```diff
@@ -0,0 +1,266 @@
...
+    localHelper.setUp();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    localHelper.tearDown();
+  }
+    
+  public void setRequestAsMentor() {
+    String userType = "Mentor";
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 29]
```diff
@@ -0,0 +1,266 @@
...
+    String mentorName = null;
+    when(request.getParameter("profName-input")).thenReturn(mentorName);
+    String company = null;
+    when(request.getParameter("company-input")).thenReturn(company);
+    String specialty = null;
+    when(request.getParameter("specialty-input")).thenReturn(specialty);
+
+    profileBuilderServlet.doPost(request, response);
+    verify(response, times(1)).sendRedirect("/logged_in_homepage.html");
+    Assert.assertEquals(0, dataService.prepare(new Query("User")).countEntities(withLimit(10)));
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:115]: Yes

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java (hunk 21 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java:214]: Yes It is. I removed both.

Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/ProfileBuilderServletTest.java [hunk 30]
```diff
@@ -0,0 +1,266 @@
...
+    when(request.getParameter("school-input")).thenReturn(school);
+    String major = "Computer Science";
+    when(request.getParameter("major-input")).thenReturn(major);
+
+    profileBuilderServlet.doPost(request, response);
+    PreparedQuery results = ds.prepare(new Query("User"));
+    String nameResult = "";
+    String schoolResult = "";
+    String majorResult = "";
+    for (Entity entity : results.asIterable()) {
```
[the remark this answers is not in the document]
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/ProfileBuilderServletTest.java]: Done

File: src/test/java/com/google/sps/PersistentUserRepositoryTest.java [hunk 31]
```diff
@@ -84,96 +84,76 @@ public void addFakeMentorsTest() {
...
-    Collection<String> specialtiesResult = new HashSet<>();
-
-    for (User allUser : allUsers) {
-        int len = allUser.getSpecialties().size();
-        if(len > 0){
-            specialtiesResult = allUser.getSpecialties(); 
-        }
-    }
-    Assert.assertEquals(expected, specialtiesResult);
+    Collection<User> allUsers = repository.getAllUsers();
```
Comment (ghandre) [inline src/test/java/com/google/sps/PersistentUserRepositoryTest.java:106]: L106: This is now unused.

File: src/main/java/com/google/sps/data/PersistentUserRepository.java [hunk 32]
```diff
@@ -157,9 +180,7 @@ public User fetchUserWithId(String userId) {
     PreparedQuery results = getQueryFilterForId(userId);
     Collection<User> userProfiles = fetchUserEntities(results);
     if(userProfiles.size() == 1){
-        for (User userProfile : userProfiles) {
-            return userProfile;
-        }
+        return userProfiles.iterator().next();
     }
     User fake = new User(userId);
```
Comment (ghandre) [inline src/main/java/com/google/sps/data/PersistentUserRepository.java:185]: What's the reasoning for returning a fake user with an id as a name?
If this is meant just to ensure you have a value to return, then it might make more sense to do something like:
`if(userProfiles.size() != 1) throw new Exception("User \"" + name + "\" does not exist");`
`return userProfiles.iterator().next()`

File: src/test/java/com/google/sps/PersistentUserRepositoryTest.java (hunk 31 above)
Comment (samuelsandoval1) [reply src/test/java/com/google/sps/PersistentUserRepositoryTest.java]: Done

File: src/main/java/com/google/sps/data/PersistentUserRepository.java (hunk 32 above)
Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java:185]: i got an error about the `return userProfiles.iterator().next()` It says its unreachable since it should check for 1 size only

Comment (samuelsandoval1) [reply src/main/java/com/google/sps/data/PersistentUserRepository.java]: Done
